### DISCUSSION TITLE
Multiplayer VR Flight Simulator made with UE5

### DISCUSSION THREAD

**Discussion by demonixis** [JSBSim-Team/jsbsim#920] [Show and tell]
Hello everyone,
Today I'm glad to share with you what I've been able to do with JSBSim and a few other tech. It's a flight simulator made for VR and multiplayer in mind (but it works in flat and single player too).

## The video
Here is [the demo video](https://www.youtube.com/watch?v=AzTUtWNRk3w&list=PLK6ml1527ppgqg61fqZR47BaTh97AIU_a)

## JSBSim
It's an heavily modified version of JSBSim that supports:
- Multiplayer (JSBSim is only active on the server side)
- Mobile (Meta Quest/Android support, still wip), there is a WIP PR for that
- Helicopters (it's more a WIP wrapper)
- A few quality of life improvements, some of them where already merged in master, other are waiting for merging, other are not yet done

## The simulator
As you can see in the video, it's made for multiplayer in mind, you can control an aircraft with two pilots at the same time. Inputs are synchronized on the server. In LAN, it works pretty well. I've not yet tested with an online connection, but I think it's better in LAN.

The main FDM is JSBSim, but there is another one FDM for some helicopters and for a more arcade way on airplanes. The FDM can be switched from the menu, during the gameplay.

Joystick, VR Controllers and game controllers are supported.

Do you have questions? I hope you'll like it :)

Yannick.

**Comment by seanmcleod** [comment]
So assuming 2 players I assume that you're running your Unreal based app on each of their PCs and then transmitting their inputs to a 3rd PC, the server PC which is running JSBSim. Which then sums the inputs ala Airbus style, feeds the sum into JSBSim and then the JSBSim pose data and other data is broadcast to the two client PCs which then use it to update the pose in each Unreal environment?

**Reply by demonixis** [reply]
This is simpler but you're right. One of the PC runs the game on host mode, that means it's the server but also a player. Other PC will connect on that host. 

The input reconciliation is done on the server, it mixes yaw, pitch and roll based on priority.

**Reply by Iskander229** [reply]
so is the replication server authoritative? I wonder how smooth it is no matter the speed of an aircraft, sadly the video attached is no longer available

**Reply by demonixis** [reply]
It was a host server and yes authoritative. I used the SmoothSync plugin for replication (extrapolation + prediction), it's was great on local network
